tuf/api/metadata.py

```diff
@@ -1331,6 +1433,25 @@ def __init__(
         self.roles = roles
         self.unrecognized_fields = unrecognized_fields or {}
 
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Delegations):
+            return False
+
+        # Check that the order of the roles is the same as dictionaries
+        # are insensitive to order in comparisons (see issue #1788).
+        if (
+            isinstance(self.roles, dict)
+            and isinstance(other.roles, dict)
+            and list(self.roles.keys()) != list(other.roles.keys())
+        ):
+            return False
```

Comment:
I assume the `list()` calls are here because dict_keys does not consider order in eq just like dict, so you need to construct a list? Annoying but makes sense.


I'm wondering about this: a few lines later we do `self.roles == other.roles` and that presumably does the first two checks above right? So if we move the list comparison after `self.roles == other.roles` then we would not need the first two checks or this if-clause at all...

I think with a good comment that would still be understandable and 5 fewer lines of code

Reply:
> `self.roles == other.roles` can return `True` when both `roles` attributes are other types

that case doesn't sound like something eq() needs to care about though?

Reply:
Based on a chat it's clear that you are worried about the code raising an error later if roles is not a dict. I don't have a very strong opinion but it seems to me that eq() should be allowed to crash if the objects internal structure is invalid -- we can't be prepared for everything and should catch issues with roles type elsewhere _if_ we really don't want that to happen. eq() implementations should be as simple as possible

Reply:
I kind of agree that we should allow it to crash.
What do you think @lukpueh ?

Reply:
I like how you placed the check for the order inside the final check in the return. 
I wanted to do something like it before but was hesitant because I wanted to have a comment, but short comment like this could do the work. :)